SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/Lobby.cs

```diff
@@ -0,0 +1,383 @@
+                public override CMsgClientMMSGetLobbyList.Filter Serialize()
+                {
+                    CMsgClientMMSGetLobbyList.Filter filter = base.Serialize();
+                    filter.value = Value.ToString();
+                    return filter;
+                }
+            }
+
+            /// <summary>
+            /// Can be used to filter lobbies by comparing a string against a value in each lobby's metadata.
+            /// </summary>
+            public sealed class StringFilter : Filter
+            {
+                /// <summary>
+                /// String value to compare against.
+                /// </summary>
+                public string Value { get; }
+
+                /// <summary>
+                /// Initializes a new instance of the <see cref="StringFilter"/> class.
+                /// </summary>
+                /// <param name="key">The metadata key this filter pertains to.</param>
+                /// <param name="comparison">The comparison method used by this filter.</param>
+                /// <param name="value">String value to compare against.</param>
+                public StringFilter( string key, ELobbyFilterComparison comparison, string value ) : base( ELobbyFilterType.String, key, comparison )
+                {
+                    Value = value;
+                }
+
+                /// <summary>
+                /// Serializes the string filter into a representation used internally by SteamMatchmaking.
+                /// </summary>
+                /// <returns>A protobuf serializable representation of this filter.</returns>
+                public override CMsgClientMMSGetLobbyList.Filter Serialize()
+                {
+                    CMsgClientMMSGetLobbyList.Filter filter = base.Serialize();
+                    filter.value = Value;
+                    return filter;
+                }
+            }
+
+            /// <summary>
+            /// Can be used to filter lobbies by comparing a string against a value in each lobby's metadata.
+            /// </summary>
+            public sealed class Member
+            {
+                /// <summary>
+                /// SteamID of the lobby member.
+                /// </summary>
+                public SteamID SteamID { get; }
+
+                /// <summary>
+                /// Steam persona of the lobby member.
+                /// </summary>
+                public string PersonaName { get; }
+
+                /// <summary>
+                /// Metadata attached to the lobby member.
+                /// </summary>
+                public IReadOnlyDictionary<string, string> Metadata { get; }
+
+                internal Member( SteamID steamId, string personaName, IReadOnlyDictionary<string, string> metadata = null )
+                {
+                    SteamID = steamId;
+                    PersonaName = personaName;
+                    Metadata = metadata ?? new Dictionary<string, string>();
```

Comment:
Can we have a single empty `Dictionary` instead of one per `Member`?

Reply:
We probably should use `AsReadOnly`, otherwise consumers can cast the object reference back to it's mutable implementation.

Reply:
> Can we have a single empty Dictionary instead of one per `Member`?

Can do. Just wasn't sure what the precedent for that was withing this code-base. Very happy to make the change as I do similar in my own code-bases.

> We probably should use `AsReadOnly`, otherwise consumers can cast the object reference back to it's mutable implementation.

I'm not typically a fan of doing of stuff like this. If the user is determined to violate our public contract they can always use reflection etc. It's a bit of fruitless battle.

Reply:
Just realised I'd missed these two pieces of feedback. Addressing both points now.